src/libappimage/desktop_integration/Thumbnailer.cpp

```diff
@@ -0,0 +1,156 @@
+            auto resources = extractResources(appImage);
+            std::string appIcon = getAppIconName(resources);
+
+            std::string canonicalPathMd5 = getCanonicalPathMd5(appImage.getPath());
+
+            auto normalIconData = getIconData(resources, appIcon, "128x128");
+            generateNormalSizeThumbnail(canonicalPathMd5, normalIconData);
+
+            auto largeIconData = getIconData(resources, appIcon, "256x256");
+            generateLargeSizeThumbnail(canonicalPathMd5, largeIconData);
+        }
+
+        void Thumbnailer::remove(const core::AppImage& appImage) {
+            std::string canonicalPathMd5 = getCanonicalPathMd5(appImage.getPath());
+            bf::path normalThumbnailPath = getNormalThumbnailPath(canonicalPathMd5);
+            bf::path largeThumbnailPath = getLargeThumbnailPath(canonicalPathMd5);
+
+            bf::remove(normalThumbnailPath);
+            bf::remove(largeThumbnailPath);
+        }
+
+        void Thumbnailer::generateNormalSizeThumbnail(const std::string& canonicalPathMd5,
+                                                      std::vector<char>& normalIconData) const {
+
+            bf::path normalThumbnailPath = getNormalThumbnailPath(canonicalPathMd5);
+            bf::create_directories(normalThumbnailPath.parent_path());
+
+            try {
+                std::vector<uint8_t> iconData{normalIconData.begin(), normalIconData.end()};
+                IconHandle iconHandle(iconData);
+                iconHandle.setSize(128);
+                iconHandle.save(normalThumbnailPath.string(), "png");
+                return;
+            } catch (const IconHandleError&) {}
+
+            // It wasn't possible to generate a thumbnail, therefore the the icon will be written as it's
+            bf::ofstream out(normalThumbnailPath);
+            out.write(normalIconData.data(), normalIconData.size());
+        }
+
+        void Thumbnailer::generateLargeSizeThumbnail(const std::string& canonicalPathMd5,
+                                                     std::vector<char>& largeIconData) const {
+            bf::path largeThumbnailPath = getLargeThumbnailPath(canonicalPathMd5);
+            bf::create_directories(largeThumbnailPath.parent_path());
+
+            try {
+                std::vector<uint8_t> iconData{largeIconData.begin(), largeIconData.end()};
+                IconHandle iconHandle(iconData);
+                iconHandle.setSize(256);
+                iconHandle.save(largeThumbnailPath.string(), "png");
+                return;
+            } catch (const IconHandleError&) {}
+
+            // It wasn't possible to generate a thumbnail, therefore the the icon will be written as it's
+            bf::ofstream out(largeThumbnailPath);
+            out.write(largeIconData.data(), largeIconData.size());
+            out.close();
+        }
+
+
+        std::string Thumbnailer::getCanonicalPathMd5(const std::string& appImagePath) const {
```

Comment:
The size of an MD5 hash is known, so this could also just return a `std::array`, which is way more efficient size wise.

Reply:
Indeed it would be more efficient but it will end converted to an `std::string` when appended to the file names so the improvement will be lost. That's why I would prefer to keep it as an `std::string`.

Reply:
The point is that it's safer to use an `std::array` here, as it will ensure the size. Internally, some char buffers are used anyway. The conversion can be done by the user, the contract of this method doesn't need to contain "returns dynamic string type".

Reply:
This function was removed as a duplicated of utils::hashPath